### PR TITLE
[MOS-744] Fix MTP integration

### PR DESCRIPTION
**Pull request by Lefucjusz**
Fixed issues with integration
of MTP with filesystem:
- fixed memory corruption in get_disk_properties();
- remove spare space hack causing MTP to return
invalid free space and
capacity of the storage;
- added mtime and ctime handling;
- minor code cleanup.

### CODE REVIEW

File: mtp/mtp_fs.cpp [hunk 1]
```diff
@@ -4,336 +4,320 @@
...
+        const struct mtp_fs *fs = static_cast<struct mtp_fs *>(arg);
+        struct statvfs stvfs {};
+        statvfs(fs->root, &stvfs);
+
+        [[maybe_unused]] const auto freeSpace = stvfs.f_bavail * stvfs.f_bsize;
+        const auto capacity                   = stvfs.f_blocks * stvfs.f_bsize;
+        disk_properties.capacity              = capacity;
+
+        log_debug("Capacity: %u MB, free: %u MB",
+                  static_cast<unsigned>(capacity / 1024 / 1024),
```
Comment (marziel) [inline mtp/mtp_fs.cpp:68]: A minor thing, but may be confusing for someone: these are de facto MiBs, not MBs.

File: mtp/mtp_fs.cpp [hunk 2]
```diff
@@ -4,336 +4,320 @@
...
-    statvfs( purefs::dir::getUserDiskPath().c_str(), &stvfs);
-
-    // TODO: stats are for entire storage. If MTP is intended to expose
-    // only one directory, these stats should be recalculated
-    const auto freeSpace = stvfs.f_bsize * stvfs.f_bavail;
-    data->freespace = (getSpareSpace() > freeSpace) ? 0 : freeSpace - getSpareSpace();
-    data->capacity =  stvfs.f_frsize * stvfs.f_blocks;
+    const mtp_storage_properties_t *get_disk_properties(void *arg)
+    {
+        const struct mtp_fs *fs = static_cast<struct mtp_fs *>(arg);
```
Comment (marziel) [inline mtp/mtp_fs.cpp:59]: As this is C++, _auto_ would make this statement much clearer to read.

File: mtp/mtp_fs.cpp [hunk 3]
```diff
@@ -4,336 +4,320 @@
...
+        struct statvfs stvfs {};
+        statvfs(fs->root, &stvfs);
```
Comment (marziel) [inline mtp/mtp_fs.cpp:61]: Shouldn't the return value be checked here?

File: mtp/mtp_fs.cpp [hunk 4]
```diff
@@ -4,336 +4,320 @@
...
-    while( (de=readdir(fs->find_data)) &&  is_dot(de->d_name)) {
-        LOG("skip: %s", de->d_name);
-    }
-    if( !de )
+    uint16_t ext_to_format_code(const char *name)
     {
-        LOG("No files found");
-        *count = 0;
-        return 0;
+        if (strstr(".jpg", name) != nullptr) {
```
Comment (marziel) [inline mtp/mtp_fs.cpp:141]: This way of checking the file extension can be unreliable, see https://github.com/mudita/MuditaOS/pull/4121#discussion_r1003080987.

File: mtp/mtp_fs.cpp [hunk 5]
```diff
@@ -4,336 +4,320 @@
...
+        }
+        new_handle = mtp_db_add(fs->db, info->filename);
+        if (new_handle == 0) {
+            log_error("Map is full. Can't create: %s", info->filename);
+            return -1;
+        }
+
+        fs->file = std::fopen(absolutePath.c_str(), "w");
+        if (fs->file == nullptr) {
+            log_error("Failed to open %s", info->filename);
```
Comment (marziel) [inline mtp/mtp_fs.cpp:223]: What with the product of the above _mtp_db_add()_, then?

File: mtp/mtp_fs.cpp (hunk 3 above)
Comment (Lefucjusz) [reply mtp/mtp_fs.cpp]: Done

File: mtp/mtp_fs.cpp (hunk 4 above)
Comment (Lefucjusz) [reply mtp/mtp_fs.cpp]: Fixed

Comment (marziel) [reply mtp/mtp_fs.cpp]: I think that .jp**e**g is also a valid extension for this format.

File: mtp/mtp_fs.cpp (hunk 5 above)
Comment (Lefucjusz) [reply mtp/mtp_fs.cpp:240]: Added removing the file from the database if `fopen` has failed

File: mtp/mtp_fs.cpp (hunk 4 above)
Comment (Lefucjusz) [reply mtp/mtp_fs.cpp]: You're right, added

File: mtp/mtp_fs.cpp (hunk 5 above)
Comment (marziel) [reply mtp/mtp_fs.cpp:246]: Now it's correct. By the way - as this is C++, you could use some RAII to simplify this resource management. If sticking to C, there's a good "goto cleanup" idiom: https://stackoverflow.com/questions/788903/valid-use-of-goto-for-error-management-in-c

**Pull request merged**